### PR TITLE
[WIP] Continuous deployment (CD) - Docker build and push to GHCR

### LINKED ISSUE #159: GH Action to build deployable docker image for `fah-xchem`, push to registry

### PR DESCRIPTION
**Pull request by dotsdl**
Closes #159.

We build a deployable image containing `fah-xchem` and all dependencies installed via `conda` to GitHub Container Registry as the primary source for fah-xchem docker images.

## Status
- [x] Ready to go

### CODE REVIEW

**Comment by dotsdl** [comment]
Success!

We have our new image pushed to the registry: https://github.com/choderalab/fah-xchem/pkgs/container/fah-xchem/versions

Can execute `fah-xchem` CLI from anywhere with, e.g.:

```
docker run ghcr.io/choderalab/fah-xchem:sha-4206c4d --help
```

**Comment by dotsdl** [comment]
Ready for review @jchodera!

File: environment.yml [hunk 1]
```diff
@@ -26,6 +26,7 @@ dependencies:
   - networkx
   - beautifulsoup4
   - pip
+  - compilers
   - pip:
       - git+https://github.com/openforcefield/Arsenic.git@0.1.0
```
Comment (jchodera) [inline environment.yml:31]: There's now an `arsenic` conda-forge package!
https://anaconda.org/conda-forge/arsenic

File: devtools/docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,20 @@
+FROM continuumio/miniconda3
+
+SHELL ["/bin/bash", "-c"]
+
+COPY . fah-xchem
+
+RUN conda env create -n fah-xchem -f fah-xchem/environment.yml \
+ && source activate fah-xchem \
```
Comment (jchodera) [inline devtools/docker/Dockerfile:8]: Should there be a `conda clean` in here to clean up all the staged packages?

File: .github/workflows/docker-build.yaml [hunk 3]
```diff
@@ -0,0 +1,56 @@
+---
+# build docker image, push to registry
+
+name: Build and Push Docker Image
```
Comment (jchodera) [inline .github/workflows/docker-build.yaml:4]: Remind me of the advantages in building here and pushing the image instead of on dockerhub?

**Review by jchodera — APPROVED** [review]
This looks good!

File: environment.yml (hunk 1 above)
Comment (dotsdl) [reply environment.yml:31]: Ah perfect! We'll switch to that.

File: devtools/docker/Dockerfile (hunk 2 above)
Comment (dotsdl) [reply devtools/docker/Dockerfile:8]: Yes, good catch. Adding that now. We certainly don't need cached, compressed versions of the packages lying about.

File: .github/workflows/docker-build.yaml (hunk 3 above)
Comment (dotsdl) [reply .github/workflows/docker-build.yaml:4]: I did a research cycle on registry options we could push to, including initially Dockerhub and AWS ECR. Dockerhub was the choice I'd reach for previously, but I still don't see a great way to push without an explicit user account, and the free tier for orgs gives us only 3 seats, all filled. A service/bot account is what I would use for this automation, with permissions scoped appropriately.

Github Container Registry in its current form [appears new as of earlier this year](https://github.blog/2021-06-21-github-packages-container-registry-generally-available/). It wasn't on my radar before, but it solves the seat issue and avoids Dockerhub's other free tier restrictions (e.g. number of repos). For open source projects it (currently) gives us everything we need for free, so seems just as good if not better a distribution hub for container images as Dockerhub would be.

**Pull request merged**